gadgets/deadlock/go/program.go

```diff
@@ -0,0 +1,231 @@
+// limitations under the License.
+
+package main
+
+import (
+	"fmt"
+	"sort"
+	"strings"
+
+	api "github.com/inspektor-gadget/inspektor-gadget/wasmapi/go"
+
+	"deadlock/digraph"
+)
+
+// The graphs map stores a directed graph for each process ID.
+var graphs map[uint32]*digraph.DiGraph
+
+// The detectedCycles map stores the set of detected cycles for each PID.
+var detectedCycles map[uint32]map[string]struct{}
+
+//export gadgetInit
+func gadgetInit() int {
+	graphs = make(map[uint32]*digraph.DiGraph)
+	detectedCycles = make(map[uint32]map[string]struct{})
+
+	// Get the mutex datasource and its fields
+	dsMutex, err := api.GetDataSource("mutex")
+	if err != nil {
+		api.Warnf("failed to get datasource: %s", err)
+		return 1
+	}
+	mutex1F, err := dsMutex.GetField("mutex1")
+	if err != nil {
+		api.Warnf("failed to get field: %s", err)
+		return 1
+	}
+	mutex2F, err := dsMutex.GetField("mutex2")
+	if err != nil {
+		api.Warnf("failed to get field: %s", err)
+		return 1
+	}
+	tidF, err := dsMutex.GetField("tid")
+	if err != nil {
+		api.Warnf("failed to get field: %s", err)
+		return 1
+	}
+	pidF, err := dsMutex.GetField("pid")
+	if err != nil {
+		api.Warnf("failed to get field: %s", err)
+		return 1
+	}
+	mutex1StackF, err := dsMutex.GetField("mutex1_stack_id")
+	if err != nil {
+		api.Warnf("failed to get field: %s", err)
+		return 1
+	}
+	mutex2StackF, err := dsMutex.GetField("mutex2_stack_id")
+	if err != nil {
+		api.Warnf("failed to get field: %s", err)
+		return 1
+	}
+	commF, err := dsMutex.GetField("comm")
+	if err != nil {
+		api.Warnf("failed to get field: %s", err)
+		return 1
+	}
+	mntnsIdF, err := dsMutex.GetField("mntns_id")
+	if err != nil {
+		api.Warnf("failed to get field: %s", err)
+		return 1
+	}
+	err = dsMutex.Unreference()
+	if err != nil {
+		api.Warnf("failed to unreference datasource: %s", err)
+		return 1
+	}
+
+	// Create the deadlock datasource and its fields
+	dsDeadlock, err := api.NewDataSource("deadlock", api.DataSourceTypeSingle)
+	if err != nil {
+		api.Warnf("failed to create datasource: %s", err)
+		return 1
+	}
+	deadlockPidF, err := dsDeadlock.AddField("pid", api.Kind_Uint32)
+	if err != nil {
+		api.Warnf("failed to add field: %s", err)
+		return 1
+	}
+	deadlockNodesF, err := dsDeadlock.AddField("nodes", api.Kind_Uint64)
+	if err != nil {
+		api.Warnf("failed to add field: %s", err)
+		return 1
+	}
+	deadlockMntnsIdF, err := dsDeadlock.AddField("mntns_id", api.Kind_Uint64)
+	if err != nil {
+		api.Warnf("failed to add field: %s", err)
+		return 1
+	}
+	err = deadlockMntnsIdF.AddTag("type:gadget_mntns_id") // enrich mntns id
+	if err != nil {
+		api.Warnf("failed to add tag to field: %s", err)
+		return 1
+	}
+
+	// Build the graph for deadlock detection using edges from the mutex datasource
+	dsMutex.SubscribeArray(func(source api.DataSource, data api.DataArray) error {
+		affectedPIDs := make(map[uint32]struct{})
+		for i := 0; i < data.Len(); i++ {
+			datum := data.Get(i)
```

Comment:
The documentation of [dataArrayGet](https://inspektor-gadget.io/docs/latest/gadget-devel/gadget-wasm-api-raw#dataarraygetd-uint32-index-uint32-uint32) says that it returns 0 on error.

At some point, I got this error in a loop:
```
WARN[0102] too many open handles                        
WARN[0102] handle 0 not found                           
WARN[0102] handle 0 not found                           
WARN[0102] handle 0 not found                           
WARN[0102] handle 0 not found                           
WARN[0102] handle 0 not found                           
WARN[0102] handle 0 not found                           
WARN[0102] handle 0 not found                           
WARN[0102] handle 0 not found                           
```

Notice that the message "handle 0 not found" is printed 8 times and the variable datum is used 8 times below, so I guess it is the source of the problem.

I think you need to:
- call `data.Release(datum)` in the for loop after you no longer need it.
- test if datum is 0 and call api.Warnf+continue if so

@flyth IMO, the wasm golang API could be improved to do that automatically:
```go
data.RunCallbackAt(i, func(datum) {
    mutex1, err := mutex1F.Uint64(datum)
    ...
})
```

Reply:
I got that error too and I am working on it. AFAIK it has nothing to do with the missing call to `.Release()`

Reply:
Opened PR for releasing the data manually in https://github.com/inspektor-gadget/inspektor-gadget/pull/3702